### PR TITLE
Fix stdClass inconsistencies when serializing to JSON

### PR DESCRIPTION
**Pull request by goetas**
Fixes https://github.com/schmittjoh/serializer/issues/61
Closes https://github.com/schmittjoh/serializer/issues/59 
Closes https://github.com/schmittjoh/serializer/issues/242

### CODE REVIEW

File: tests/JMS/Serializer/Tests/Serializer/JsonSerializationTest.php [hunk 1]
```diff
@@ -263,7 +265,85 @@ public function testSerializeWithNonUtf8EncodingWhenDisplayErrorsOn()
 
     public function testSerializeArrayWithEmptyObject()
     {
-        $this->assertEquals('{"0":{}}', $this->serialize(array(new \stdClass())));
+        $this->assertEquals('[{}]', $this->serialize(array(new \stdClass())));
```
Comment (goetas) [inline tests/JMS/Serializer/Tests/Serializer/JsonSerializationTest.php]: this to me sounds a wrong test case.

**Comment by goetas** [comment]
|version|time|
---|---
|master|0.044855713844299|
|pull request|0.038354706764221|

This PR gives also a 15% of speed improvement

File: tests/JMS/Serializer/Tests/Serializer/JsonSerializationTest.php (hunk 1 above)
Comment (goetas) [inline tests/JMS/Serializer/Tests/Serializer/JsonSerializationTest.php]: @schmittjoh What do you think?

**Comment by robwasripped** [comment]
With the class `GenericSerializationVisitor` now deprecated, what is the recommended way to ensure that the visitor given in the `serializer.post_serialize` event contains the `setData` method?

**Comment by goetas** [comment]
I really do not suggest to use `setData`. when possible is better IMO to use `vistiProperty(new StaticMedadataProperty(....))`. 

But to answer to your question, using directly `JsonSerializationVisitor` should work well. 
(BTW with the previous implementation, `setData` was fully relying on the internal `json_encode` logic)

**Pull request merged**